tools/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/ActiveTransactionRegistry.java

```diff
@@ -79,39 +79,36 @@ private ActiveTransactionRegistry() {
 			}
 		}
 
-		primaryCache = CacheBuilder.newBuilder().removalListener(new RemovalListener<UUID, Transaction>() {
-
-			@Override
-			public void onRemoval(RemovalNotification<UUID, Transaction> notification) {
-				Transaction entry = notification.getValue();
-				try {
-					entry.close();
-				} catch (RepositoryException | InterruptedException | ExecutionException e) {
-					// fall through
-				}
-			}
-		}).build();
-
-		secondaryCache = CacheBuilder.newBuilder().removalListener(new RemovalListener<UUID, Transaction>() {
-
-			@Override
-			public void onRemoval(RemovalNotification<UUID, Transaction> notification) {
-				if (RemovalCause.EXPIRED.equals(notification.getCause())) {
-					final UUID transactionId = notification.getKey();
-					final Transaction entry = notification.getValue();
-					synchronized (primaryCache) {
-						if (!entry.hasActiveOperations()) {
-							// no operation active, we can decommission this entry
-							primaryCache.invalidate(transactionId);
-							logger.warn("deregistered expired transaction {}", transactionId);
-						} else {
-							// operation still active. Reinsert in secondary cache.
-							secondaryCache.put(transactionId, entry);
+		primaryCache = CacheBuilder.newBuilder()
+				.removalListener((RemovalNotification<UUID, Transaction> notification) -> {
+					Transaction entry = notification.getValue();
+					try {
+						entry.close();
+					} catch (RepositoryException | InterruptedException | ExecutionException e) {
+						// fall through
+					}
+				})
+				.build();
+
+		secondaryCache = CacheBuilder.newBuilder()
+				.removalListener((RemovalNotification<UUID, Transaction> notification) -> {
+					if (RemovalCause.EXPIRED.equals(notification.getCause())) {
+						final UUID transactionId = notification.getKey();
+						final Transaction entry = notification.getValue();
+						synchronized (primaryCache) {
+							if (!entry.hasActiveOperations()) {
+								// no operation active, we can decommission this entry
+								primaryCache.invalidate(transactionId);
+								logger.warn("deregistered expired transaction {}", transactionId);
+							} else {
+								// operation still active. Reinsert in secondary cache.
+								secondaryCache.put(transactionId, entry);
+							}
```

Comment:
Interesting: the Jenkins build fails on this, saying that (on line 105) `secondaryCache` may not have been initialized.

Reply:
indeed, maybe a if not null may solve it, I'll look into it

Reply:
It’s a scoping issue. Lambdas expect all variables in their scope to effectively be final. Because the secondaryCache is bound after the scope has been determined for the lambda, the lambda assumes that’s its uninitialised. 

Can you attach the removal listener after build() has bee called?

Also, I think whole approach is a bit funky. Is there no place to attach a willSoonBeRemovedListener instead?

Reply:
There seems to be no method to attach the listener after building the cache... a workaround seems to be to create a getter, and use that inside the lambda instead of the variable.

Another dirty way would be to just surpress the warning, but IMHO that isn't a good practice...

Reply:
I like the setter approach.

For the wider issue of the design of this cache, I'd happily revisit when we have time. I remember designing it this way because I was somewhat hampered by the limitations of the Guava cache.